# HG changeset patch
# User Rik <rik@octave.org>
# Date 1499292959 25200
#      Wed Jul 05 15:15:59 2017 -0700
# Node ID 4c048a2792bc0dd8105fa0554a9ddbbeb7f2680b
# Parent  45a6347fdd848554ae25ec85fd76468b6d5982b5
Use frexp from std library rather than gnulib.

* bootstrap.conf: Remove modules frexp, frexpf.

* liboctave/wrappers/math-wrappers.c, liboctave/wrappers/math-wrappers.h:
Remove wrapper files for gnulib frexp implementation.

* liboctave/wrappers/module.mk: Remove wrapper files from build system.

* lo-mappers.h (log2, frexp): Replace 2-input form of log2, and frexp, function
prototypes with inline calls to std::frexp.

* lo-mappers.cc: Remove #include for "math-wrappers.h".
Remove 2-input form of log2, and also frexp implementations.

diff --git a/bootstrap.conf b/bootstrap.conf
--- a/bootstrap.conf
+++ b/bootstrap.conf
@@ -33,18 +33,16 @@ gnulib_modules="
   crypto/sha1
   crypto/sha256
   crypto/sha512
   dup2
   fcntl
   filemode
   fnmatch
   fpucw
-  frexp
-  frexpf
   fseek
   ftell
   ftruncate
   getcwd
   gethostname
   getopt-gnu
   getrusage
   gettimeofday
diff --git a/liboctave/numeric/lo-mappers.cc b/liboctave/numeric/lo-mappers.cc
--- a/liboctave/numeric/lo-mappers.cc
+++ b/liboctave/numeric/lo-mappers.cc
@@ -22,17 +22,16 @@ along with Octave; see the file COPYING.
 */
 
 #if defined (HAVE_CONFIG_H)
 #  include "config.h"
 #endif
 
 #include "lo-mappers.h"
 #include "lo-specfun.h"
-#include "math-wrappers.h"
 
 // FIXME: We used to have this situation:
 //
 //   Functions that forward to gnulib belong here so we can keep
 //   gnulib:: out of lo-mappers.h.
 //
 // but now we just use std:: and explicit wrappers in C++ code so maybe
 // some of the forwarding functions can be defined inline here.
@@ -118,50 +117,28 @@ namespace octave
       FloatComplex y = std::asin (x);
 
       if (std::imag (x) == 0.0f && std::real (x) > 1.0f)
         return std::conj (y);
       else
         return y;
     }
 
-    double frexp (double x, int *expptr)
-    {
-      return octave_frexp_wrapper (x, expptr);
-    }
-
-    float frexp (float x, int *expptr)
-    {
-      return octave_frexpf_wrapper (x, expptr);
-    }
-
     Complex
     log2 (const Complex& x)
     {
       return std::log (x) / M_LN2;
     }
 
     FloatComplex
     log2 (const FloatComplex& x)
     {
       return std::log (x) / static_cast<float> (M_LN2);
     }
 
-    double
-    log2 (double x, int& exp)
-    {
-      return frexp (x, &exp);
-    }
-
-    float
-    log2 (float x, int& exp)
-    {
-      return frexp (x, &exp);
-    }
-
     Complex
     log2 (const Complex& x, int& exp)
     {
       double ax = std::abs (x);
       double lax = log2 (ax, exp);
       return (ax != lax) ? (x / ax) * lax : x;
     }
 
diff --git a/liboctave/numeric/lo-mappers.h b/liboctave/numeric/lo-mappers.h
--- a/liboctave/numeric/lo-mappers.h
+++ b/liboctave/numeric/lo-mappers.h
@@ -93,18 +93,24 @@ namespace octave
     }
 
     inline double log2 (double x) { return std::log2 (x); }
     inline float log2 (float x) { return std::log2f (x); }
 
     extern OCTAVE_API Complex log2 (const Complex& x);
     extern OCTAVE_API FloatComplex log2 (const FloatComplex& x);
 
-    extern OCTAVE_API double log2 (double x, int& exp);
-    extern OCTAVE_API float log2 (float x, int& exp);
+    inline double log2 (double x, int& exp)
+    {
+      return std::frexp (x, &exp);
+    }
+    inline float log2 (float x, int& exp)
+    {
+      return std::frexp (x, &exp);
+    }
 
     extern OCTAVE_API Complex log2 (const Complex& x, int& exp);
     extern OCTAVE_API FloatComplex log2 (const FloatComplex& x, int& exp);
 
     inline double exp2 (double x) { return std::exp2 (x); }
     inline float exp2 (float x) { return std::exp2f (x); }
 
     template <typename T>
@@ -176,18 +182,25 @@ namespace octave
 
     template <typename T>
     std::complex<T>
     roundb (const std::complex<T>& x)
     {
       return std::complex<T> (roundb (std::real (x)), roundb (std::imag (x)));
     }
 
-    extern OCTAVE_API double frexp (double x, int *expptr);
-    extern OCTAVE_API float frexp (float x, int *expptr);
+    inline double frexp (double x, int *expptr)
+    {
+      return std::frexp (x, expptr);
+    }
+
+    inline float frexp (float x, int *expptr)
+    {
+      return std::frexp (x, expptr);
+    }
 
     inline bool isnan (bool) { return false; }
     inline bool isnan (char) { return false; }
 
     inline bool isnan (double x) { return std::isnan (x); }
     inline bool isnan (float x) { return std::isnan (x); }
 
     // FIXME: Do we need the isnan overload for complex?
diff --git a/liboctave/wrappers/math-wrappers.c b/liboctave/wrappers/math-wrappers.c
deleted file mode 100644
--- a/liboctave/wrappers/math-wrappers.c
+++ /dev/null
@@ -1,46 +0,0 @@
-/*
-
-Copyright (C) 2016-2017 Mike Miller
-
-This file is part of Octave.
-
-Octave is free software; you can redistribute it and/or modify it
-under the terms of the GNU General Public License as published by
-the Free Software Foundation; either version 3 of the License, or
-(at your option) any later version.
-
-Octave is distributed in the hope that it will be useful, but
-WITHOUT ANY WARRANTY; without even the implied warranty of
-MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU General Public License for more details.
-
-You should have received a copy of the GNU General Public License
-along with Octave; see the file COPYING.  If not, see
-<http://www.gnu.org/licenses/>.
-
-*/
-
-// These functions may be provided by gnulib.  We don't include gnulib
-// headers directly in Octave's C++ source files to avoid problems that
-// may be caused by the way that gnulib overrides standard library
-// functions.
-
-#if defined (HAVE_CONFIG_H)
-#  include "config.h"
-#endif
-
-#include <math.h>
-
-#include "math-wrappers.h"
-
-double
-octave_frexp_wrapper (double x, int *expptr)
-{
-  return frexp (x, expptr);
-}
-
-float
-octave_frexpf_wrapper (float x, int *expptr)
-{
-  return frexpf (x, expptr);
-}
diff --git a/liboctave/wrappers/math-wrappers.h b/liboctave/wrappers/math-wrappers.h
deleted file mode 100644
--- a/liboctave/wrappers/math-wrappers.h
+++ /dev/null
@@ -1,40 +0,0 @@
-/*
-
-Copyright (C) 2016-2017 Mike Miller
-
-This file is part of Octave.
-
-Octave is free software; you can redistribute it and/or modify it
-under the terms of the GNU General Public License as published by
-the Free Software Foundation; either version 3 of the License, or
-(at your option) any later version.
-
-Octave is distributed in the hope that it will be useful, but
-WITHOUT ANY WARRANTY; without even the implied warranty of
-MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU General Public License for more details.
-
-You should have received a copy of the GNU General Public License
-along with Octave; see the file COPYING.  If not, see
-<http://www.gnu.org/licenses/>.
-
-*/
-
-#if ! defined (octave_math_wrappers_h)
-#define octave_math_wrappers_h 1
-
-#if defined __cplusplus
-extern "C" {
-#endif
-
-double
-octave_frexp_wrapper (double x, int *expptr);
-
-float
-octave_frexpf_wrapper (float x, int *expptr);
-
-#if defined __cplusplus
-}
-#endif
-
-#endif
diff --git a/liboctave/wrappers/module.mk b/liboctave/wrappers/module.mk
--- a/liboctave/wrappers/module.mk
+++ b/liboctave/wrappers/module.mk
@@ -7,17 +7,16 @@ NOINSTALL_WRAPPERS_INC = \
   %reldir%/fcntl-wrappers.h \
   %reldir%/filepos-wrappers.h \
   %reldir%/fpucw-wrappers.h \
   %reldir%/gen-tempname-wrapper.h \
   %reldir%/getopt-wrapper.h \
   %reldir%/glob-wrappers.h \
   %reldir%/hash-wrappers.h \
   %reldir%/localcharset-wrapper.h \
-  %reldir%/math-wrappers.h \
   %reldir%/mkostemp-wrapper.h \
   %reldir%/nanosleep-wrapper.h \
   %reldir%/nproc-wrapper.h \
   %reldir%/octave-popen2.h \
   %reldir%/putenv-wrapper.h \
   %reldir%/set-program-name-wrapper.h \
   %reldir%/signal-wrappers.h \
   %reldir%/stat-wrappers.h \
@@ -44,17 +43,16 @@ WRAPPERS_SRC = \
   %reldir%/fcntl-wrappers.c \
   %reldir%/filepos-wrappers.c \
   %reldir%/fpucw-wrappers.c \
   %reldir%/gen-tempname-wrapper.c \
   %reldir%/getopt-wrapper.c \
   %reldir%/glob-wrappers.c \
   %reldir%/hash-wrappers.c \
   %reldir%/localcharset-wrapper.c \
-  %reldir%/math-wrappers.c \
   %reldir%/mkostemp-wrapper.c \
   %reldir%/nanosleep-wrapper.c \
   %reldir%/nproc-wrapper.c \
   %reldir%/octave-popen2.c \
   %reldir%/putenv-wrapper.c \
   %reldir%/set-program-name-wrapper.c \
   %reldir%/signal-wrappers.c \
   %reldir%/stat-wrappers.c \
